File: lib-segment/src/main/java/websiteschema/mpsegment/filter/EnglishStemFilter.java
package websiteschema.mpsegment.filter;

import websiteschema.mpsegment.lang.en.PorterStemmer;
import websiteschema.mpsegment.util.CharCheckUtil;

/**
 * EnglishStemFilter should not be used in search engine. <br/>
 * Because this filter will delete white space characters,<br/>
 * which many search engines already took it into account.
 */
public class EnglishStemFilter extends AbstractSegmentFilter {
    PorterStemmer porterStemmer = new PorterStemmer();

    @Override
    public void doFilter() {
        int length = segmentResult.length();
        for (int index = 0; index < length; index++) {
            String word = segmentResult.getWord(index);
            if (CharCheckUtil.isEnglish(word)) {
                segmentResult.setWord(index, porterStemmer.stem(word));
                deleteNextSpaceWord(index, length);
            }
        }
    }

    private void deleteNextSpaceWord(int index, int length) {
        for (int i = index + 1; i < length; i++) {
            if (CharCheckUtil.isWhiteSpace(segmentResult.getWord(i))) {
                deleteWordAt(i);
            } else {
                break;
            }
        }
    }
}

File: lib-segment-scala/src/main/scala/websiteschema/mpsegment/filter/EnglishStemFilter.scala
package websiteschema.mpsegment.filter

import websiteschema.mpsegment.lang.en.PorterStemmer
import websiteschema.mpsegment.util.CharCheckUtil

class EnglishStemFilter(isStemmingWord: Boolean) extends AbstractSegmentFilter {
  val porterStemmer = new PorterStemmer()

  override def doFilter() {
    val length = segmentResult.length()
    for (index <- 0 until length) {
      val word = segmentResult.getWord(index)
      if (CharCheckUtil.isEnglish(word)) {
        if (isStemmingWord) {
          segmentResult.setWord(index, porterStemmer.stem(word))
        }
        deleteNextSpaceWord(index, length)
      }
    }
  }

  private def deleteNextSpaceWord(index: Int, length: Int) {
    for (i <- index + 1 until length) {
      if (CharCheckUtil.isWhiteSpace(segmentResult.getWord(i))) {
        deleteWordAt(i)
      } else {
        return
      }
    }
  }
}
